### ISSUE TITLE
Performance decrease a lot during redis cluster fail over period 

### ISSUE THREAD

**Issue by Neway6655** [xetorthio/jedis#1512]
### Expected behavior
I have a Redis Cluster with 5 master nodes and 5 slave nodes, each master has one slave. 
I have running a stress test by calling jedis's hmget method, then I killed one master, so redis cluster would fail over, and my expectation is the performance might be decreased a little during fail over period, since some keys might hit the failing master node. And it will recover after redis cluster finish master/slave switch.

### Actual behavior
But actually, performance was decreased a lot during the fail over period, about 75 percentage, the performance was down to 1/4 as before. And after redis cluster recovered, the performance recovered as well.

### Steps to reproduce:
1. start redis cluster.
2. start a stress test.
3. kill one redis master node.

#### Jedis version:
2.8.2
#### Redis version:
3.0.5
#### Java version:
JDK 1.7u21

**Comment by Neway6655** [comment]
I found jedis will try to renewSlotCache when JedisConnectionException happened, while redis cluster fail over may need some time (several seconds or minutes), code snippet as below:

```
   catch (JedisConnectionException jce) {
      // release current connection before recursion
      releaseConnection(connection);
      connection = null;

      if (attempts <= 1) {
        //We need this because if node is not reachable anymore - we need to finally initiate slots renewing,
        //or we can stuck with cluster state without one node in opposite case.
        //But now if maxAttempts = 1 or 2 we will do it too often. For each time-outed request.
        //TODO make tracking of successful/unsuccessful operations for node - do renewing only
        //if there were no successful responses from this node last few seconds
        this.connectionHandler.renewSlotCache();

        //no more redirections left, throw original exception, not JedisClusterMaxRedirectionsException, because it's not MOVED situation
        throw jce;
      }

      return runWithRetries(key, attempts - 1, tryRandomNode, asking);
    }
```

so during the fail over, requests hit at the failing node will fail into JedisConnectionException, then renewSlotCache will be called continuously, this may decrease the performance a lot since renewSlotCache will cause write lock called, which will lock the read lock in getConnectionFromSlot method.

**Comment by marcosnils** [comment]
@Neway6655 good catch. Is there something that you propose to overcome this situation?. We haven't been actively developing Jedis for while, so we really depend on our users to help us with their problems and to propose a possible solution. 

Also @HeartSaVioR or @Spikhalskiy  might have a word regarding this.

**Comment by kenny319** [comment]
@marcosnils We are trying to reduce the impaction to the normal nodes when the failover is in progress. The core principle is to narrow the use of the write lock in JedisClusterInfoCache. 
1. Make sure only one thread can retrieve the new slots by using another lock
2. Compare the new slots when old slots and replace the old slots with new slots when they are different only avoiding unnecessary renew operation.

**Comment by mohammadagha** [comment]
Hey @Neway6655  can you share your stress test script ?

**Comment by Neway6655** [comment]
@mohammadagha Actually I found this issue during the robustness test of my application, so the script is my application's robustness test script, it's not suitable for others.

**Comment by AnujTewari** [comment]
@Neway6655 I ma facing the issue (using version 3.2.0) and came across this issue which is still open
How did you deal with this issue?
cc @marcosnils

**Comment by Neway6655** [comment]
@AnujTewari I think you can check @kenny319's commit: https://github.com/kenny319/jedis/commit/2b722e0ad05c8b9ee774a3479b8f1aec68a8080c

which we used as a hot fix in our environments.
